### PR TITLE
[SPARK-48898][SQL] Set nullability correctly in the Variant schema

### PR DESCRIPTION
**Pull request by cashmand**
### What changes were proposed in this pull request?

The `variantShreddingSchema` method converts a human-readable schema for Variant to one that's a valid shredding schema. According to the shredding schema in https://github.com/apache/parquet-format/pull/461, each shredded field in an object should be a required group - i.e. a non-nullable struct. This PR fixes the `variantShreddingSchema` to mark that struct as non-nullable.

### Why are the changes needed?

If we use `variantShreddingSchema` to construct a schema for Parquet, the schema would be technically non-conformant with the spec by setting the group as optional. I don't think this should really matter to readers, but it would waste a bit of space in the Parquet file by adding an extra definition level.

### Does this PR introduce _any_ user-facing change?

No, this code is not used yet.

### How was this patch tested?

Added a test to do some minimal validation of the `variantShreddingSchema` function.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Review by chenhao-db — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/VariantWriteShreddingSuite.scala [hunk 1]
```diff
@@ -67,6 +67,34 @@ class VariantWriteShreddingSuite extends SparkFunSuite with ExpressionEvalHelper
 
   private val emptyMetadata: Array[Byte] = parseJson("null").getMetadata
 
+  test("variantShreddingSchema") {
+    // Validate the schema produced by SparkShreddingUtils.variantShreddingSchema for a few simple
+    // cases.
+    // metadata is always non-nullable.
+    assert(SparkShreddingUtils.variantShreddingSchema(IntegerType) ==
+        StructType(Seq(StructField("metadata", BinaryType, nullable = false),
+                       StructField("value", BinaryType, nullable = true),
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/VariantWriteShreddingSuite.scala:76]: can we have double spaced indentation?

Comment (cashmand) [reply sql/core/src/test/scala/org/apache/spark/sql/VariantWriteShreddingSuite.scala]: Done.

**Review by HyukjinKwon — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by dongjoon-hyun** [comment]
Interestingly, it seems to fail in `master` branch somehow. Could you take a look at the failures, please, @cashmand and all?

Screenshot 2024-12-11 at 10 23 41

- https://github.com/apache/spark/actions/runs/12280389975/job/34266758469
- https://github.com/apache/spark/actions/runs/12280542565/job/34267268862

```
[info] *** 1 TEST FAILED ***
[error] Failed: Total 12430, Failed 1, Errors 0, Passed 12429, Ignored 33, Canceled 1
[error] Failed tests:
[error] 	org.apache.spark.sql.VariantShreddingSuite
[error] (sql / Test / test) sbt.TestsFailedException: Tests unsuccessful
```

**Comment by dongjoon-hyun** [comment]
It seems that the following landed faster and this PR didn't test the test case (which fail currently)
- https://github.com/apache/spark/pull/48851

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/SparkShreddingUtils.scala [hunk 2]
```diff
@@ -61,8 +61,11 @@ case object SparkShreddingUtils {
           StructField(TypedValueFieldName, arrayShreddingSchema, nullable = true)
         )
       case StructType(fields) =>
+        // The field name level is always non-nullable: Variant null values are represented in the
+        // "value" columna as "00", and missing values are represented by setting both "value" and
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/SparkShreddingUtils.scala:65]: `columna` -> `column`?

**Comment by dongjoon-hyun** [comment]
Sorry but let me revert this to recover the CI. Although it looks like a trivial Mal-formed error case testing, could you re-submit this PR to make it sure to pass all?

**Comment by cashmand** [comment]
> Sorry but let me revert this to recover the CI and unblock other PRs. Although it looks like a trivial Mal-formed error case testing, could you re-submit this PR to make it sure to pass all?

Hi @dongjoon-hyun, I opened https://github.com/apache/spark/pull/49151, which is exactly the same, but updates the broken test. Can you please take a look, and merge if it looks okay?

**Comment by dongjoon-hyun** [comment]
Thank you so much, @cashmand . If CI passes, I'll merge back ASAP.

**Pull request closed without merging**